core/trino-main/src/main/java/io/trino/server/security/oauth2/OAuth2Authenticator.java

```diff
@@ -36,23 +41,33 @@
 {
     private final OAuth2Service service;
     private final String principalField;
+    private final Optional<String> groupsField;
+    private final UserMapping userMapping;
 
     @Inject
     public OAuth2Authenticator(OAuth2Service service, OAuth2Config config)
     {
-        super(createUserMapping(config.getUserMappingPattern(), config.getUserMappingFile()));
         this.service = requireNonNull(service, "service is null");
         this.principalField = config.getPrincipalField();
+        groupsField = requireNonNull(config.getGroupsField(), "groupsField is null");
+        userMapping = createUserMapping(config.getUserMappingPattern(), config.getUserMappingFile());
     }
 
     @Override
-    protected Optional<Principal> extractPrincipalFromToken(String token)
+    protected Optional<Identity> createIdentity(String token)
+            throws UserMappingException
     {
         try {
-            return service.convertTokenToClaims(token)
-                    .map(claims -> claims.get(principalField))
-                    .map(String.class::cast)
-                    .map(BasicPrincipal::new);
+            Optional<Map<String, Object>> claims = service.convertTokenToClaims(token);
+            if (claims.isEmpty()) {
+                return Optional.empty();
+            }
+            String principal = (String) claims.get().get(principalField);
+            Identity.Builder identity = Identity.forUser(userMapping.mapUser(principal));
+            identity.withPrincipal(new BasicPrincipal(principal));
+            groupsField.flatMap(field -> Optional.ofNullable((List<String>) claims.get().get(field)))
```

Comment:
No need for `flatMap` as `map` handles `null` correctly.

Reply:
`Optional.map(o -> Optional.of(o))` returns `Optional<Optional<>>` 
whereas
`Optional.flatMap(o -> Optional.of(o))` returns `Optional<>`